core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/ProcessUtil.java

```diff
@@ -126,6 +129,73 @@ public static Process executeAndWait(
     return process;
   }
 
+  /**
+   * Start the process, writing the stdout of process to {@code stdout}, stderr of process to {@code
+   * stderr}, and terminate process by {@code timeout}.<br>
+   *
+   * @param commandLine arguments of process command
+   * @param timeout timeout for process. If process duration > {@code timeout} than kill process and
+   *     throw {@link TimeoutException}.
+   * @param timeUnit timeUnit of the {@code timeout}.
+   * @param stdout a consumer where stdout will be redirected
+   * @param stderr a consumer where stderr will be redirected
+   * @return the started process
+   * @throws InterruptedException in case terminate process
+   * @throws IOException in case I/O error, or command execution error, or stderr is not empty after
+   *     the command execution
+   * @throws TimeoutException if process gets more time then defined by {@code timeout}
+   */
+  public static Process executeAndWait(
+      String[] commandLine,
+      int timeout,
+      TimeUnit timeUnit,
+      ListLineConsumer stdout,
+      ListLineConsumer stderr)
+      throws TimeoutException, IOException, InterruptedException {
+    ProcessBuilder pb = new ProcessBuilder(commandLine);
+
+    Process process = pb.start();
+
+    CompletableFuture<Void> future =
+        CompletableFuture.runAsync(
+            () -> {
+              try {
+                // consume logs until process ends
+                readOutput(process, stdout, stderr);
+              } catch (IOException e) {
+                LOG.error(
+                    format(
+                        "Failed to complete reading of the process '%s' output due to occurred error",
+                        Joiner.on(" ").join(commandLine)),
+                    e);
+              }
+            });
+
+    if (!process.waitFor(timeout, timeUnit)) {
+      try {
+        ProcessUtil.kill(process);
+      } catch (RuntimeException x) {
+        LOG.error("An error occurred while killing process '{}'", Joiner.on(" ").join(commandLine));
+      }
+      throw new TimeoutException(
+          format(
+              "Process '%s' was terminated by timeout %s %s.",
+              Joiner.on(" ").join(commandLine), timeout, timeUnit.name().toLowerCase()));
+    }
+
+    future.join(); // wait until process output is read
+
+    String errorMessage = stderr.getText();
+    if (!errorMessage.isEmpty()) {
```

Comment:
Some processes write into stderr. 
You'd have better to check result code.

Reply:
Good catch. Will check if `process.exitCode()` is suitable for the case of execution of docker command which fails.

Reply:
what will happen if stderr contains 2GB of data? 10GB of data?

Reply:
@riuvshin said that from the point of selenium tests workspace logs hardly ever can have size more than hundred Mb, so we may not to check the size additionally.

Reply:
Good catch. We can limit size of text which we are putting into the exception, say, to hundreds symbols.

Reply:
@tolusha:
> Some processes write into stderr.
> You'd have better to check result code.

In that case `process#start()` throws **IOException**, and for that case we have already had another method [org.eclipse.che.api.core.util.ProcessUtil#executeAndWait(java.lang.String[], int, java.util.concurrent.TimeUnit, org.eclipse.che.api.core.util.LineConsumer)](https://github.com/eclipse/che/blob/b90cc2dfd86af12ea680f9ce6177410a5f946224/core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/ProcessUtil.java#L96).

at the same time proposed method is dedicated to handle the case when exit code = 0, but **stderr** is not empty - that is what we sometimes have as a result of docker command execution.

Reply:
I have found out that `ProcessBuilder#start()` doesn't throw **IOException** in case of running error-prone command as following: `bash -c '<error-prone_command>` despite of non-zero exit code. So, additional check of non-zero exit code will be enough.